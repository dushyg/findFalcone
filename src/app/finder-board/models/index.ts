export { IFalconAppState } from './falconApp.state';
export { RawPlanet, IPlanet } from './planet';
export { RawVehicle, IVehicle } from './vehicle';
export { default as PlanetChange } from './planetChange';
export { default as VehicleChange } from './vehicleChange';
export { ISearchAttempt } from './searchAttempt';
