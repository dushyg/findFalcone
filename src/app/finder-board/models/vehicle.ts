// tslint:disable: indent
export interface IVehicle {
  name: string;
  totalNumUnits: number;
  maxDistance: number;
  speed: number;
  availNumUnits: number;
}

export interface RawVehicle {
  id?: number;
  name: string;
  total_no: number;
  max_distance: number;
  speed: number;
}
