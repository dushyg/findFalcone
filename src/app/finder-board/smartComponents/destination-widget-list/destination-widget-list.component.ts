import { Component, OnInit } from "@angular/core";
import { FinderFacadeService } from "src/app/finder-board/services/finder-facade.service";

@Component({
  selector: "app-destination-widget-list",
  templateUrl: "./destination-widget-list.component.html",
  styleUrls: ["./destination-widget-list.component.scss"],
})
export class DestinationWidgetListComponent implements OnInit {
  constructor(private finderFacadeService: FinderFacadeService) {}

  public widgetCountIterator: number[];

  ngOnInit() {
    this.widgetCountIterator = new Array<number>(
      this.finderFacadeService.getCountOfWidgetsDisplayed()
    ).fill(0);
  }
}
