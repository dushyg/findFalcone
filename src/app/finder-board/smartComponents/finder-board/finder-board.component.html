<div
  class="boardContainer"
  *ngIf="finderFacadeService.dashboardVm$ | async as vm"
>
  <div>
    <div class="timeTaken">Time Taken : {{ vm.timeTaken }}</div>

    <div class="instructionLine">
      Select planets you want to search in :
    </div>

    <app-destination-widget-list
      [maxWidgetCount]="maxWidgetCount"
    ></app-destination-widget-list>

    <div class="findButtonContainer">
      <input
        [disabled]="!vm.isReadyForSearch"
        type="button"
        value="Find Falcone!"
        (click)="findFalcone()"
      />
    </div>

    <div class="d-flex justify-content-center my-spinner">
      <div
        class="fade show"
        style="width: 15rem; height: 15rem;"
        *ngIf="vm.isLoading"
      >
        <span>Loading...</span>
        <div
          class="spinner-border m-5"
          style="width: 5rem; height: 5rem;"
          role="status"
        ></div>
      </div>
    </div>
  </div>
</div>
