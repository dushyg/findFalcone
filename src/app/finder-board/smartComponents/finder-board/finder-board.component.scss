div.boardContainer {
  /*default css for small devices like IPhone 5*/
  text-align: center;
  height: 100%;
  display: grid;
  grid-template-areas:
    'timeTaken'
    'instructionLine'
    'widgetList'
    'findButton';
  grid-template-columns: 100%;
  grid-template-rows: 40px 40px 1fr 50px;

  div.instructionLine {
    grid-area: instructionLine;
  }

  div.timeTaken {
    grid-area: timeTaken;
    margin-top: 10px;
  }

  div.instructionLine {
    font-size: 0.5rem;
  }

  app-destination-widget-list {
    grid-area: widgetList;
    overflow-y: auto;
  }

  div.findButtonContainer {
    grid-area: findButton;
  }

  div.instructionLine,
  app-destination-widget-list,
  div.findButtonContainer {
    margin-bottom: 10px;
  }

  app-destination-widget-list {
    min-height: 320px;
  }

  @media (min-width: 375px) {
    /*css for small devices like IPhone 6/7*/
    grid-template-rows: 45px 45px 1fr 50px;
    div.instructionLine {
      font-size: 0.65rem;
    }
  }

  @media (min-width: 768px) {
    /*css for medium devices like IPad*/
    grid-template-rows: 50px 50px 1fr 50px;
    div.instructionLine {
      font-size: 0.9rem;
    }
  }

  @media (min-width: 1240px) {
    /*css for large devices like IPad Pro and larger*/
    div.instructionLine {
      font-size: 1rem;
    }
  }
}
