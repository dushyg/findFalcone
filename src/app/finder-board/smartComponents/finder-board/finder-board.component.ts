import {
  Component,
  OnInit,
  OnDestroy,
  ChangeDetectionStrategy,
} from '@angular/core';
import { FinderFacadeService } from 'src/app/finder-board/services/finder-facade.service';
import { Router } from '@angular/router';
import { takeWhile } from 'rxjs/operators';

@Component({
  selector: 'app-finder-board',
  templateUrl: './finder-board.component.html',
  styleUrls: ['./finder-board.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FinderBoardComponent implements OnInit, OnDestroy {
  constructor(
    public finderFacadeService: FinderFacadeService,
    private router: Router
  ) {}

  public maxWidgetCount = 0;
  ngOnInit() {
    this.finderFacadeService.initializeAppData();
    this.maxWidgetCount = this.finderFacadeService.getCountOfWidgetsDisplayed();
  }

  public findFalcone() {
    this.router.navigate(['result']);
  }

  ngOnDestroy(): void {}
}
