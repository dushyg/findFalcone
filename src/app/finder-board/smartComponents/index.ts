export { DestinationWidgetComponent } from './destination-widget/destination-widget.component';
export { FinderBoardComponent } from './finder-board/finder-board.component';
export { VehicleListComponent } from './vehicle-list/vehicle-list.component';
