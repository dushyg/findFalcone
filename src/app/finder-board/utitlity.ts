export const createSpyObj = (methodNames): { [key: string]: jest.Mock<any> } => {
    const obj: any = {};

    for (let i = 0; i < methodNames.length; i++) {
        obj[methodNames[i]] = jest.fn();
    }

    return obj;
};
