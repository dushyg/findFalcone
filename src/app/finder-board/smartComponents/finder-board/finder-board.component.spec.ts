import {
  async,
  ComponentFixture,
  TestBed,
  fakeAsync,
  tick,
  flush,
} from '@angular/core/testing';

import { FinderBoardComponent } from './finder-board.component';
import { IPlanet } from 'src/app/finder-board/models/planet';
import { IVehicle } from 'src/app/finder-board/models/vehicle';
import { createSpyObj } from 'src/app/finder-board/utitlity';
import { PlanetsService } from 'src/app/finder-board/services/planets.service';
import { VehiclesService } from 'src/app/finder-board/services/vehicles.service';
import { FinderFacadeService } from 'src/app/finder-board/services/finder-facade.service';
import { Router } from '@angular/router';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { FalconeTokenService } from '../../services/falcone-token.service';
import { By } from '@angular/platform-browser';
import { of, asyncScheduler } from 'rxjs';
import { DestinationWidgetListComponent } from '../../presenterComponents/destination-widget-list/destination-widget-list.component';

describe('FinderBoardComponent', () => {
  let component: FinderBoardComponent;
  let fixture: ComponentFixture<FinderBoardComponent>;

  let planetServiceMock: { [key: string]: jest.Mock<any, any> };
  let vehiclesServiceMock: { [key: string]: jest.Mock<any, any> };
  let apiTokenServiceMock: { [key: string]: jest.Mock<any, any> };
  let routerServiceMock: { [key: string]: jest.Mock<any, any> };
  let planetListToBeReturned: IPlanet[];
  let vehicleListToBeReturned: IVehicle[];
  const apiTokenToBeReturned = { token: 'plmVHX' };
  beforeEach(() => {
    planetServiceMock = createSpyObj(['getAllPlanets']);
    vehiclesServiceMock = createSpyObj(['getAllVehicles']);
    apiTokenServiceMock = createSpyObj(['getFalconeFinderApiToken']);
    routerServiceMock = createSpyObj(['navigate']);

    TestBed.configureTestingModule({
      declarations: [FinderBoardComponent, DestinationWidgetListComponent],
      providers: [
        FinderFacadeService,
        { provide: PlanetsService, useValue: planetServiceMock },
        { provide: VehiclesService, useValue: vehiclesServiceMock },
        { provide: FalconeTokenService, useValue: apiTokenServiceMock },
        { provide: Router, useValue: routerServiceMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();

    planetListToBeReturned = [
      { name: 'Donlon', distance: 100 },
      { name: 'Enchai', distance: 200 },
      { name: 'Jebing', distance: 300 },
      { name: 'Sapir', distance: 400 },
    ];

    vehicleListToBeReturned = [
      {
        name: 'Space pod',
        availNumUnits: 2,
        maxDistance: 200,
        speed: 2,
        totalNumUnits: 2,
      },
      {
        name: 'Space rocket',
        availNumUnits: 1,
        maxDistance: 300,
        speed: 4,
        totalNumUnits: 1,
      },
      {
        name: 'Space shuttle',
        availNumUnits: 1,
        maxDistance: 400,
        speed: 5,
        totalNumUnits: 1,
      },
      {
        name: 'Space ship',
        availNumUnits: 2,
        maxDistance: 600,
        speed: 10,
        totalNumUnits: 2,
      },
    ];

    fixture = TestBed.createComponent(FinderBoardComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should render time taken with initial value of 0', () => {
    planetServiceMock.getAllPlanets.mockReturnValue(planetListToBeReturned);
    vehiclesServiceMock.getAllVehicles.mockReturnValue(vehicleListToBeReturned);
    apiTokenServiceMock.getFalconeFinderApiToken.mockReturnValue(
      apiTokenToBeReturned
    );
    fixture.detectChanges();
    const divTimeTaken = fixture.debugElement.query(By.css('.timeTaken'))
      .nativeElement as HTMLElement;
    expect(divTimeTaken).toBeTruthy();
    expect(divTimeTaken.textContent.includes('Time Taken : 0'));
  });

  it('should render app-destination-widget-list', () => {
    planetServiceMock.getAllPlanets.mockReturnValue(planetListToBeReturned);
    vehiclesServiceMock.getAllVehicles.mockReturnValue(vehicleListToBeReturned);
    apiTokenServiceMock.getFalconeFinderApiToken.mockReturnValue(
      apiTokenToBeReturned
    );

    fixture.detectChanges();

    const widgetList = fixture.debugElement.query(
      By.directive(DestinationWidgetListComponent)
    );

    expect(widgetList).toBeTruthy();
    expect(
      (widgetList.componentInstance as DestinationWidgetListComponent)
        .widgetCountIterator.length
    ).toBe(4);
  });

  it('should render disabled find button', () => {
    planetServiceMock.getAllPlanets.mockReturnValue(
      of(planetListToBeReturned, asyncScheduler)
    );
    vehiclesServiceMock.getAllVehicles.mockReturnValue(
      of(vehicleListToBeReturned, asyncScheduler)
    );
    apiTokenServiceMock.getFalconeFinderApiToken.mockReturnValue(
      of(apiTokenToBeReturned, asyncScheduler)
    );
    fixture.detectChanges();
    const findButton = fixture.debugElement.query(
      By.css('div.findButtonContainer input[type=button]:disabled')
    );
    expect(findButton).toBeTruthy();
  });
});
