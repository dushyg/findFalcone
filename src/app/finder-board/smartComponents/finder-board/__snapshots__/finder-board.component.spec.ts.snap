// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinderBoardComponent should create 1`] = `
<app-finder-board
  finderFacadeService={[Function FinderFacadeService]}
  maxWidgetCount={[Function Number]}
  router={[Function Object]}
>
  <div
    class="boardContainer"
  >
    <div>
      <div
        class="timeTaken"
      >
        Time Taken : 0
      </div>
      <div
        class="instructionLine"
      >
         Select planets you want to search in : 
      </div>
      <app-destination-widget-list>
        <div
          class="widgetListContainer"
        >
          
          <div
            class="widgetRow"
          >
            <app-destination-widget />
          </div>
          <div
            class="widgetRow"
          >
            <app-destination-widget />
          </div>
          <div
            class="widgetRow"
          >
            <app-destination-widget />
          </div>
          <div
            class="widgetRow"
          >
            <app-destination-widget />
          </div>
        </div>
      </app-destination-widget-list>
      <div
        class="findButtonContainer"
      >
        <input
          disabled=""
          type="button"
          value="Find Falcone!"
        />
      </div>
      <div
        class="d-flex justify-content-center my-spinner"
      >
        
      </div>
    </div>
  </div>
</app-finder-board>
`;
